src/LegalizeVectors.cpp

```diff
@@ -0,0 +1,630 @@
+            return result;
+        }
+        return IRMutator::visit(op);
+    }
+
+    Expr visit(const VectorReduce *op) override {
+        if (max_lanes == 0) {
+            return IRMutator::visit(op);
+        }
+        const Expr &arg = op->value;
+        if (arg.type().lanes() > max_lanes) {
+            int vecs_per_reduction = op->value.type().lanes() / op->type.lanes();
+            if (vecs_per_reduction % max_lanes == 0) {
+                // This should be possible too. TODO
+            }
+
+            internal_assert(op->type.lanes() == 1) << "Vector legalization currently does not support VectorReduce with lanes != 1: " << Expr(op);
+            int num_vecs = (arg.type().lanes() + max_lanes - 1) / max_lanes;
+            Expr result;
+            for (int i = 0; i < num_vecs; i++) {
+                int lane_start = i * max_lanes;
+                int lane_count_for_vec = std::min(arg.type().lanes() - lane_start, max_lanes);
+                Expr partial_arg = mutate(ExtractLanes(lane_start, lane_count_for_vec, max_lanes).mutate(arg));
+                Expr partial_red = VectorReduce::make(op->op, std::move(partial_arg), op->type.lanes());
+                if (i == 0) {
+                    result = partial_red;
+                } else {
+                    switch (op->op) {
+                    case VectorReduce::Add:
+                        result = result + partial_red;
+                        break;
+                    case VectorReduce::SaturatingAdd:
+                        result = saturating_add(result, partial_red);
+                        break;
+                    case VectorReduce::Mul:
+                        result = result * partial_red;
+                        break;
+                    case VectorReduce::Min:
+                        result = min(result, partial_red);
+                        break;
+                    case VectorReduce::Max:
+                        result = max(result, partial_red);
+                        break;
+                    case VectorReduce::And:
+                        result = result && partial_red;
+                        break;
+                    case VectorReduce::Or:
+                        result = result || partial_red;
+                        break;
+                    }
+                }
+            }
+            return result;
+        } else {
+            return IRMutator::visit(op);
+        }
+    }
+};
+
+}  // namespace
+
+Stmt legalize_vectors(const Stmt &s) {
+    // Similar to CSE, lifting out stuff into variables.
+    // Pass 1): lift out Shuffles that exceed lane count into variables
+    // Pass 2): Rewrite those vector variables as bundles of vector variables, while legalizing all other stuff.
+    Stmt m0 = simplify(s);
+    Stmt m1 = common_subexpression_elimination(m0, false);
+    if (!m1.same_as(s)) {
+        debug(3) << "After CSE:\n"
+                 << m1 << "\n";
+    }
+    Stmt m2 = LiftLetToLetStmt().mutate(m1);
```

Comment:
Unconditionally turning lets into letstmts for all backends scares me a little. I'm worried about cases where, e.g. codegen needs to do analysis on the index of a store, and it would have succeeded if the index had the let statements it depends on included in the Expr, but because we're already inside the letstmts, we don't have access to them. In theory analysis like this should all take a scope of enclosing stuff, but I don't know if that actually happens in practice. So in principle this change should be fine, but it substantially raises the testing bar we need. It might be safer to have an outer mutator that only runs this logic on the bodies of for loops with a device api that has a vector lane limit. It would make compilation faster for code not using such device apis anyway.

Reply:
Does this concern apply, given that this legalization pass is pretty much in the very back of the lowering process? I'd assume that all these kinds of analysis are already done by that point?

Reply:
There are still some things that happen in codegen. E.g. when lowering integer division to LLVM IR, Halide attempts to prove the denominator is non-zero, and if so emits simpler LLVM IR.

Reply:
I have seen this type of warning before (but not while working on the legalization, iirc).

The lifting of let to letstmt was mostly needed to handle shuffles of vectors with funny indices. The constituent vectors would become defined in LetStmts. As such, the original shuffle expression, which is legalized by breaking it up into multiple shuffles, can then reuse those constituent vectors for every generated shuffle, after which they get simplified where possible.

I'm wondering how this is different from CSE? Wouldn't generic CSE also make it in the same way impossible to prove certain things?

Reply:
CSE does sometimes cause problems of this nature, yeah. In principle this transformation should be fine. But my high-level point here is that if we were sure we're not changing the IR at all for LLVM targets like say ARM or hexagon DSP, the PR can be merged without risk of e.g. silently regressing some obscure corner case inside Google due to IR being shuffled around in a way that makes some optimization no longer trigger. It would also save compile time in the common case to only apply this pass to device API loops that need it.

Reply:
I would run the entire current body of legalize_vectors only on the affected loops, by adding a new mutator outside of this. I.e. instead of your current:

```
Stmt legalize_vectors(const Stmt &s) { ... logic ...}
```

I'd do:

```
Stmt legalize_vectors(const Stmt &s) {
  class LegalizeAllDeviceLoops : public IRMutator {
    using IRMutator::visit;
    Stmt visit(const For *op) override {
      if (op->device_api has a lane limit) {
        ... logic ...
      } else {
        return IRMutator::visit(op);
      }
    }
  } mutator;
  return mutator.mutate(s);
}
```

Reply:
Aha, that seems like it's going to reduce some repetitive code of checking if we are in such a loop. Will do that tomorrow.